docs/proposals/ingress-l4.md

```diff
@@ -0,0 +1,320 @@
+
+* [keepalived-vip](https://github.com/kubernetes/contrib/tree/master/keepalived-vip)
+* NGINX Ingress Controller
+  [nginx-ingress](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx)
+
+However these implementations rely on a ConfigMap being populated with a
+mapping of destination endpoint to service.  A single common interface for
+describing all Ingress would is highly desirable.
+
+
+## Design Considerations
+
+### State of L4 Protocol Support in Cloud Providers
+
+L4 support is not uniform across all Cloud Providers, specifically TLS
+termination is lacking in several Cloud Providers.
+
+|           | TCP | UDP | TLS |
+| --------- | --- | --- | --- |
+| AWS       |  x  |     |  x  |
+| GCE       |  x  |  x  | (1) |
+| Mesos     |  ?  |  ?  |  ?  |
+| Openstack |  x  |  x  |  x  |
+| Rackspace |  x  |  x  |     |
+| vSphere   |  x  |  x  |  x  |
+
+(1) GCE has TLS as an alpha service
+
+### TCP support
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+```
+
+While HTTP rules have host specifiers, in TCP rules hostnames would not be
+as hostname handling is specific to L5 or higher layer protocol being used over
+TCP.  TCP as a protocol specifier should be agnostic to higher layer protocols.
+
+### Ingress type specifier
+
+The current Ingress API assumes a protocol based on the presence or absence
+of a key (tls, http).  This leaves open the possibility of a spec that defines
+all protocols.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 3306
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+```
+
+A specification like the one above is arguably too complex and should be
+written as separate rules for each protocol being supported.  However, the
+loose syntax of the specification allows construction of a complex
+specification.  Using a 'type' or 'protocol' specifier would remove this
+ambiguity in the specification.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - protocol: tcp
+      port: 3306
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+
+    - protocol: tls
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+```
+
+### L4 TLS Termination
+
+While it would be very useful to be able to express TLS protected L4 Ingress,
```

Comment:
I dislike not having a way to express this.  Having Ingress be complete LCD across all providers I think is a flaw, because it will encourage others to end up building higher level Ingress rules that solve the higher level reqts.  I think we need to simply have a case that some cloud providers can't provide certain features.  Anyone who needs TLS terminated TCP shouldn't be using a provider that doesn't support it.